File: users-service/src/main/java/net/plshark/users/repo/GroupsRepository.java
package net.plshark.users.repo;

import net.plshark.users.model.Group;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * Repository for groups
 */
public interface GroupsRepository {

    /**
     * Get a group by ID
     * @param id the group ID
     * @return the matching group if found
     */
    Mono<Group> getForId(long id);

    /**
     * Get a group by name
     * @param name the group name
     * @return the matching group if found
     */
    Mono<Group> getForName(String name);

    /**
     * Get all groups up to a maximum number of results
     * @param maxResults the max results to return
     * @param offset the offset to start at
     * @return the groups
     */
    Flux<Group> getGroups(int maxResults, long offset);

    /**
     * Save a new group
     * @param group the group to save
     * @return the saved group
     */
    Mono<Group> insert(Group group);

    /**
     * Delete a group by ID
     * @param groupId the group ID
     * @return when complete
     */
    Mono<Void> delete(long groupId);
}

File: user-error-server/src/main/kotlin/net/plshark/users/repo/GroupsRepository.kt
package net.plshark.users.repo

import kotlinx.coroutines.flow.Flow
import net.plshark.users.model.Group
import net.plshark.users.model.GroupCreate

/**
 * Repository for groups
 */
interface GroupsRepository {

    /**
     * Find a group by ID
     * @param id the group ID
     * @return the matching group or null if not found
     */
    suspend fun findById(id: Long): Group?

    /**
     * Find a group by name
     * @param name the group name
     * @return the matching group or null if not found
     */
    suspend fun findByName(name: String): Group?

    /**
     * Find all groups up to a maximum number of results
     * @param limit the max results to return
     * @param offset the offset to start at
     * @return a [Flow] emitting the groups
     */
    fun getGroups(limit: Int, offset: Int): Flow<Group>

    /**
     * Save a new group
     * @param group the group to save
     * @return the saved group
     */
    suspend fun insert(group: GroupCreate): Group

    /**
     * Delete a group by ID
     * @param groupId the group ID
     */
    suspend fun deleteById(groupId: Long)
}
